trace-forward/test/Test/Trace/Forward/Demo/Tests.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Test.Trace.Forward.Demo.Tests
+  ( tests
+  ) where
+
+import           Control.Concurrent (forkIO, threadDelay)
+import           Control.Concurrent.STM (atomically)
+import           Control.Concurrent.STM.TBQueue (newTBQueueIO, writeTBQueue)
+import           Control.Monad (void)
+import           Data.Functor ((<&>))
+import           Data.IORef (IORef, atomicModifyIORef', newIORef, readIORef)
+import           System.Directory (getTemporaryDirectory)
+#if defined(mingw32_HOST_OS)
+import           System.FilePath ((</>), dropDrive)
+import qualified Data.Text as T
+#else
+import           System.FilePath ((</>))
+#endif
+import           Data.Time.Calendar (fromGregorian)
+import           Data.Time.Clock (UTCTime (..))
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+import           Ouroboros.Network.IOManager (withIOManager)
+
+import           Trace.Forward.Acceptor
+import           Trace.Forward.Configuration
+import           Trace.Forward.Forwarder
+import           Trace.Forward.Protocol.Type
+
+import           Test.Trace.Forward.Demo.Configs
+import           Test.Trace.Forward.Protocol.Codec ()
+import           Test.Trace.Forward.Protocol.TraceItem
+
+data Endpoint = Remote | Local
+
+tests :: TestTree
+tests = localOption (QuickCheckTests 1) $ testGroup "Trace.Forward.Demo"
+  [ testProperty "RemoteSocket" $ prop_RemoteSocket 100 Remote
+  , testProperty "LocalPipe"    $ prop_RemoteSocket 100 Local
+  ]
+
+prop_RemoteSocket :: Int -> Endpoint -> Property
+prop_RemoteSocket n ep' = ioProperty . withIOManager $ \iomgr -> do
+  ep <- case ep' of
+          Remote -> return $ RemoteSocket "127.0.0.1" 3030
+          Local  -> LocalPipe <$> mkLocalPipePath
+
+  forwarderQueue <- newTBQueueIO $ fromIntegral n
+  acceptedItems :: IORef [TraceItem] <- newIORef []
+  weAreDone <- newIORef False
+  -- Run the acceptor. It will ask 'TraceItem's from the forwarder
+  -- and store them in 'acceptorQueue'.
+  void . forkIO $
+    runTraceAcceptor
+      iomgr
+      (mkAcceptorConfig ep weAreDone)
+      (traceItemsHandler acceptedItems)
+      nodeInfoHandler
+  threadDelay 1000000
+  -- Run the forwarder. It will take 'TraceItem's from 'forwarderQueue'.
+  void . forkIO $
```

Comment:
We're leaking a thread here.  You won't notice this in `IO` but when you'll start using `IOSim` for testing you could be hit by this.

Reply:
Fixed.